test/view/set_difference.cpp

```diff
@@ -0,0 +1,317 @@
+        CONCEPT_ASSERT(Same<range_rvalue_reference_t<R2>, const int&&> ());
+    }
+
+
+    // test different orderings
+    {
+        auto res = view::set_difference(view::reverse(i1_finite), view::reverse(i2_finite), [](int a, int b)
+        {
+            return a > b;
+        });
+        ::check_equal(res, {4, 4, 3, 3, 3, 2, 1});
+    }
+
+
+    // test projections and sets with different element types
+    struct S
+    {
+        int val;
+        bool operator==(const S& other) const
+        {
+            return val == other.val;
+        }
+    };
+
+    S s_finite[] = {S{-20}, S{-10}, S{1}, S{3}, S{3}, S{6}, S{8}, S{20}};
+    
+    {
+    //     auto res1 = view::set_difference(s_finite, view::ints(-2, 10),
+    //                                      ordered_less(),
+    //                                      &S::val,
+    //                                      ident()
+    //                                     );
+        // not using &S::val avoids a problem with older libc++ where mem_fn()
+        // does not provide an operator()const
+        auto res1 = view::set_difference(s_finite, view::ints(-2, 10),
+                                         ordered_less(),
+                                         [](const S& x){ return x.val; },
+                                         ident()
+                                        );
+        using R1 = decltype(res1);
+        CONCEPT_ASSERT(Same<range_value_t<R1>, S>());
+        CONCEPT_ASSERT(Same<range_reference_t<R1>, S&>());
+        CONCEPT_ASSERT(Same<range_rvalue_reference_t<R1>, S&&> ());
+        ::check_equal(res1, {S{-20}, S{-10}, S{3}, S{20}});
+
+
+        auto res2 = view::set_difference(view::ints(-2, 10), s_finite,
+                                         ordered_less(),
+                                         ident(),
+                                         [](const S& x){ return x.val; }
+                                        );
+        using R2 = decltype(res2);
+        CONCEPT_ASSERT(Same<range_value_t<R2>, int>());
+        CONCEPT_ASSERT(Same<range_reference_t<R2>, int>());
+        CONCEPT_ASSERT(Same<range_rvalue_reference_t<R2>, int> ());
+        ::check_equal(res2, {-2, -1, 0, 2, 4, 5, 7, 9});
+    }
+    
+    
+    // move
+    {
+        auto v0 = to_<std::vector<MoveOnlyString>>({"a","b","b","c","x","x"});
+        auto v1 = to_<std::vector<MoveOnlyString>>({"b","x","y","z"});
+        auto res = view::set_difference(v0, v1, [](const MoveOnlyString& a, const MoveOnlyString& b){return a<b;});
+
+        std::vector<MoveOnlyString> expected;
+        move(res, back_inserter(expected));
+
+        ::check_equal(expected, {"a","b","c","x"});
+        ::check_equal(v1, {"b","x","y","z"});
+        //::check_equal(v0, {"","","b","","","x"});
+        
+        // FIXME: why is v0 not equal to {"","","b","","","x"}?
+        // the lazy view::set_difference mimics the greedy set_difference very closely!
```

Comment:
Here I was expressing my surprise that `set_difference` moves out the last occurences of the repeating element, as opposed to the `set_intersection`.

In hindsight, `set_difference` must find the first mismatch and can't go backwards to move out some previous instance of a repeating element. So I take back my surprised comments.

As far as I can tell `std::set_difference` only defines the number of times an equivalent element is copied out instead of which ones exactly.

Finally, `view::set_difference` and `view::set_symmetric_difference` closely emulate their eager counterpart algorithms. The lazy and eager results are verifiably the same regardless of my wrong and misinformed hunches on what an algorithm should or should not do.